### PR TITLE
Factor self attention out of the bert package

### PR DESCRIPTION
**Pull request by danieldk**
## Description

The same mechanism is reusable between transformer models. While at it, add some docstrings and make the attention mask optional.

### Types of change

Refactor

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/models/attention.py [hunk 1]
```diff
@@ -44,15 +44,24 @@ def __init__(self, *, dropout_prob: float = 0.1):
...
         """
+        Apply attention layer to the given key, query and value.
+
+        Sequence elements that are marked with `False` in the attention mask
+        are ignored by the attention mechanism (if a mask is provided).
+
+        k (Tensor): key.
+        q (Tensor): query.
+        v (Tensor): value.
+        attn_mask (Optional[AttentionMask]): attention mask.
```
Comment (shadeMe) [inline curated_transformers/models/attention.py:58]:
```suggestion
        attn_mask (Optional[AttentionMask]): attention mask.
        
```

File: curated_transformers/models/attention.py [hunk 2]
```diff
@@ -61,12 +70,112 @@ def forward(
...
+        dropout_prob: float = 0.1,
+        hidden_width: int = 768,
+        num_attention_heads: int = 12,
+    ):
+        """Construct a self-attention layer.
+
+        dropout_prob (float): dropout to apply between the self-attention
+            and output layers (default: 0.1).
+        hidden_width (int): hidden width of the layer (default: 768).
+        num_attention_heads (int): number of attenion heads (default: 12).
```
Comment (shadeMe) [inline curated_transformers/models/attention.py:101]:
```suggestion
        num_attention_heads (int): number of attention heads (default: 12).
```

File: curated_transformers/models/attention.py [hunk 3]
```diff
@@ -61,12 +70,112 @@ def forward(
...
 
         attn_weights = attn_scores.softmax(dim=-1)
         attn_values = self.dropout(attn_weights @ v)
 
         return attn_values
+
+
+# https://www.tensorflow.org/text/tutorials/transformer#multi-head_attention
+class SelfAttention(Module):
+    """Transformer self-attention layer (Vaswani et al., 2016)."""
```
Comment (shadeMe) [inline curated_transformers/models/attention.py:87]:
```suggestion
    """Transformer self-attention layer (Vaswani et al., 2017)."""
```

File: curated_transformers/models/attention.py [hunk 4]
```diff
@@ -61,12 +70,112 @@ def forward(
...
+        super().__init__()
+
+        self.model_dim = hidden_width
+        self.num_heads = num_attention_heads
+        if self.model_dim % self.num_heads != 0:
+            raise ValueError(
+                Errors.E003.format(
+                    hidden_width=self.model_dim, num_heads=self.num_heads
+                )
+            )
```
Comment (shadeMe) [inline curated_transformers/models/attention.py:112]: Shall we inline this error here already?

File: curated_transformers/models/attention.py [hunk 5]
```diff
@@ -44,15 +44,24 @@ def __init__(self, *, dropout_prob: float = 0.1):
...
+        Sequence elements that are marked with `False` in the attention mask
+        are ignored by the attention mechanism (if a mask is provided).
+
+        k (Tensor): key.
+        q (Tensor): query.
+        v (Tensor): value.
+        attn_mask (Optional[AttentionMask]): attention mask.
         Shapes:
             k, q, v - (batch, heads, seq_len, width)
             attn_mask - (batch, seq_len)
```
Comment (shadeMe) [inline curated_transformers/models/attention.py:61]: I think it makes sense to also document the output shape (in all modules)?

File: curated_transformers/models/attention.py [hunk 6]
```diff
@@ -61,12 +70,112 @@ def forward(
...
+        x (Tensor): input to apply self-attention to.
+        attn_mask (Optional[AttentionMask]): attention mask.
+
+        Shapes:
+            x - (batch, seq_len, width)
+            attn_mask - (batch, seq_len)
+        """
+
+        proj = self.input(x)
+        q, k, v = proj.chunk(3, dim=-1)
```
Comment (shadeMe) [inline curated_transformers/models/attention.py:135]: Perhaps a comment here to explain why we're doing the projection just once.

File: curated_transformers/models/attention.py [hunk 7]
```diff
@@ -61,12 +70,112 @@ def forward(
...
+    """
+    Combine the attention head representations. The inverse of
+    'split_heads'.
+
+    Shapes:
+        x - (batch, head, seq_len, width_per_head)
+        output - (batch, seq_len, width)
+    """
+    batch_size, head, seq_len, model_dim = x.size()
+    return x.transpose(1, 2).contiguous().view(batch_size, seq_len, head * model_dim)
```
Comment (shadeMe) [inline curated_transformers/models/attention.py:159]: This reads better if it came after `split_heads`, given that it's the inverse of the former.

File: curated_transformers/models/attention.py (hunk 4 above)
Comment (danieldk) [reply curated_transformers/models/attention.py]: Inlined.

File: curated_transformers/models/attention.py (hunk 5 above)
Comment (danieldk) [reply curated_transformers/models/attention.py:60]: Done for the attention module.

File: curated_transformers/models/attention.py (hunk 6 above)
Comment (danieldk) [reply curated_transformers/models/attention.py:138]: Done.

File: curated_transformers/models/attention.py (hunk 7 above)
Comment (danieldk) [reply curated_transformers/models/attention.py]: Indeed, fixed!

**Pull request merged**